main/util/src/mill/util/CoursierSupport.scala

```diff
@@ -63,136 +76,132 @@ trait CoursierSupport {
-        import scala.concurrent.ExecutionContext.Implicits.global
-        val loadedArtifacts = Gather[Task].gather(
-          for (a <- artifacts)
-            yield coursierCache.file(a).run.map(a.optional -> _)
-        ).unsafeRun()
-
-        val errors = loadedArtifacts.collect {
-          case (false, Left(x)) => x
-          case (true, Left(x)) if !x.notFound => x
-        }
-        val successes = loadedArtifacts.collect { case (_, Right(x)) => x }
 
-        (errors, successes)
-      }
+    resolutionRes.flatMap { resolution =>
+      val coursierCache0 = coursierCache(ctx, coursierCacheCustomizer)
 
-      val sourceOrJar =
-        if (sources) {
-          resolution.artifacts(
-            types = Set(coursier.Type.source, coursier.Type.javaSource),
-            classifiers = Some(Seq(coursier.Classifier("sources")))
+      val artifactsResultOrError = Artifacts(coursierCache0)
+        .withResolution(resolution)
+        .withClassifiers(
+          if (sources) Set(Classifier("sources"))
+          else Set.empty
+        )
+        .eitherResult()
+
+      artifactsResultOrError match {
+        case Left(error: DownloadingArtifacts) =>
+          val errorDetails = error.errors
+            .map(_._2)
+            .map(e => s"${System.lineSeparator()}  ${e.describe}")
+            .mkString
+          Result.Failure(
+            s"Failed to load ${if (sources) "source " else ""}dependencies" + errorDetails
           )
-        } else resolution.artifacts(
-          types = Set(
-            coursier.Type.jar,
-            coursier.Type.testJar,
-            coursier.Type.bundle,
-            coursier.Type("orbit"),
-            coursier.Type("eclipse-plugin"),
-            coursier.Type("maven-plugin")
+        case Left(error) =>
+          Result.Exception(error, new Result.OuterStack((new Exception).getStackTrace))
+        case Right(res) =>
+          Result.Success(
+            Agg.from(
+              res.files
+                .map(os.Path(_))
+                .filter(path => path.ext == "jar" && resolveFilter(path))
+                .map(PathRef(_, quick = true))
+            ) ++ localTestDeps.flatten
           )
-        )
-
-      val (errors, successes) = load(sourceOrJar)
-
-      if (errors.isEmpty) {
-        Result.Success(
-          Agg.from(
-            successes
-              .map(os.Path(_))
-              .filter(path => path.ext == "jar" && resolveFilter(path))
-              .map(PathRef(_, quick = true))
-          ) ++ localTestDeps.flatten
-        )
-      } else {
-        val errorDetails = errors.map(e => s"${System.lineSeparator()}  ${e.describe}").mkString
-        Result.Failure(
-          s"Failed to load ${if (sources) "source " else ""}dependencies" + errorDetails
-        )
       }
     }
   }
 
+  @deprecated(
+    "Prefer resolveDependenciesMetadataSafe instead, that handles errors better",
```

Comment:
```suggestion
    "Prefer resolveDependenciesMetadataSafe instead, that handles errors better",
    "Prefer resolveDependenciesMetadataSafe instead, which returns a Result instead of throwing exceptions",
```